services/lobby.service.js

```diff
@@ -0,0 +1,1003 @@
+const { Lobby, User, Question } = require("../models");
+const { lobbyHelper } = require("../helpers");
+const Constants = require('../utils/constants');
+
+const usedCodes = new Set();
+
+// create a lobby
+const createLobby = async (req, res) => {
+    try {
+        const { body } = req;
+
+        // Create a new user
+        const user = await User.create(body);
+        if (user) {
+            // Generate a unique lobby code
+            let lobbyCode = await lobbyHelper.generateUniqueLobbyCode();
+
+            // Check if the code is unique
+            while (usedCodes.has(lobbyCode)) {
+                lobbyCode = await lobbyHelper.generateUniqueLobbyCode();
+            }
+
+            // Store the code in the used codes set
+            usedCodes.add(lobbyCode);
+
+            // Create a lobby associated with the user
+            const lobbyData = {
+                lobbyCode,
+                lobbyCreator: user._id,
+                playerList: [user._id]
+            };
+
+            // Create the lobby
+            const lobby = await Lobby.create(lobbyData);
+
+            // Populate the lobby with associated user and player data
+            await lobby.populate(['lobbyCreator', 'playerList']);
+            return lobby;
+        }
+
+    } catch (error) {
+        console.log(error);
+        throw error;
+    }
+};
+
+// join lobby with lobby code
+const joinLobby = async (req, res) => {
+    try {
+        const { lobbyCode, playerName } = req.body;
+
+        // Find the lobby with the given lobbyCode
+        let findLobby = await Lobby.findOne({ lobbyCode: lobbyCode });
+
+        if (findLobby) {
+
+            // Check if the lobby is already locked
+            if (findLobby.lobbyLocked) {
+                return Constants.LOBBY_LOCKED;
+            }
+
+            // Generate a random name for the user
+            // const randomName = await lobbyHelper.generateRandomName();
+
+            // Create a new user with the random name
+            const user = await User.create({ name: playerName });
+
+            // Add the new user to the playerList array in the lobby
+            const newUserList = [...findLobby.playerList, user._id];
+
+            // Update the lobby's playerList with the new user
+            const updatedLobby = await Lobby.findByIdAndUpdate(
```

Comment:
@komalsh09 I think we will run into race conditions here while trying to update the DB, e.g. 

t=1, lobby=[A]
req=joinLobby(playerName=B) #concurrent 
req=joinLobby(playerName=C) #concurrent /

Both requests will try to update lobby=[A] and the lobby state might end up as either [A,B] or [A,C], while we were expecting it to be [A, B, C]. Our use case requires atomicity in transactions, so you will probably need to make all DB calls with [Transactions](https://www.mongodb.com/docs/manual/core/transactions/).